python/entry_point/generator.py

```diff
@@ -0,0 +1,144 @@
+#!/usr/bin/env python3
+# Copyright 2023 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""entry_point generator."""
+
+from __future__ import annotations
+
+import argparse
+import configparser
+import pathlib
+import sys
+import textwrap
+
+_ENTRY_POINTS_TXT = "entry_points.txt"
+_TEMPLATE = """\
+{shebang}
+import sys
+# Drop the first entry in the sys.path, because it will point to our workspace
+# where we are generating the entry_point script and it seems that some package
+# break due to this reason (e.g. pylint). This means that we should not use this
+# script to ever generate entry points for scripts within the main workspace,
+# but that is fine, we can create a separate generator or a boolean flag for
+# that.
+if ".runfiles" not in sys.path[0]:
+    sys.path = sys.path[1:]
+
+try:
+    from {module} import {attr}
+except ImportError:
+    entries = "\\n".join(sys.path)
+    print("Printing sys.path entries for easier debugging:")
+    print(f"sys.path is:\\n{{entries}}")
+
+if __name__ == "__main__":
+    sys.exit({entry_point}())
+"""
+
+
+class EntryPointsParser(configparser.ConfigParser):
+    """A class handling entry_points.txt
+
+    See https://packaging.python.org/en/latest/specifications/entry-points/
+    """
+
+    optionxform = staticmethod(str)
+
+
+def run(
+    *,
+    entry_points: pathlib.Path,
+    out: pathlib.Path,
+    console_script: str,
+    shebang: str,
+):
+    """Run the generator
+
+    Args:
+        entry_points: The entry_points.txt file to be parsed.
+        out: The output file.
+        console_script: The console_script entry in the entry_points.txt file.
+        shebang: The shebang to use for the console_script.
+    """
+    config = EntryPointsParser()
+    config.read(entry_points)
+    try:
+        console_scripts = dict(config["console_scripts"])
+    except KeyError:
+        raise RuntimeError(
+            f"The package does not provide any console_scripts in it's {_ENTRY_POINTS_TXT}"
+        )
+        return 1
+
+    if len(console_scripts) > 1 and not console_script:
```

Comment:
IMHO, this logic would be a bit cleaner if written in a "console script first" manner, but it's up to you.

```
if len(console_scripts) == 0:
  fail("no console scripts")
elif console_script:
  entry_point = console_scripts[console_script] or fail('{console script} not found, {keys} available')
elif len(console_scripts) == 1:
  entry_point =  console_script keys [0]
else:
  fail("must specify console script, available are {keys}")
```